torch/csrc/jit/pickler.cpp

```diff
@@ -1,54 +1,81 @@
-  if (str == "TensorID") {
-    return PicklerClass::TENSOR;
-  } else if (str == "IntList") {
-    return PicklerClass::INTLIST;
-  }
-  AT_ERROR("Unknown class name for unpickler: ", str);
-}
+const static std::unordered_map<std::string, PicklerClass> name_to_class{
+  {"TensorID", PicklerClass::TENSOR},
+  {"IntList", PicklerClass::INTLIST},
+};
 
-const std::string& getClassName(PicklerClass cls) {
-  static const std::string tensor_class("TensorID\n");
-  static const std::string intlist_class("IntList\n");
-  switch (cls) {
-    case PicklerClass::TENSOR:
-      return tensor_class;
-    case PicklerClass::INTLIST:
-      return intlist_class;
-    default:
-      AT_ERROR("Unknown class for pickler");
-  }
-}
+const static std::unordered_map<PicklerClass, std::string, std::hash<uint8_t>>
+    class_to_name{
+        {PicklerClass::TENSOR, "TensorID\n"},
+        {PicklerClass::INTLIST, "IntList\n"},
+    };
 
-const std::string& getModuleName() {
-  static const std::string module_name("__main__\n");
-  return module_name;
-}
+static std::string module_name = "__main__\n";
+
+// Protocol 2 is the highest that can be decoded by Python 2
+// See https://docs.python.org/3/library/pickle.html#data-stream-format
+constexpr static uint8_t PROTOCOL_VERSION = 2;
 
 const std::vector<char>& Pickler::stack() {
   return stack_;
 }
 
-void Pickler::start() {
+void Pickler::start(bool wrap_in_list) {
   push<OpCode>(OpCode::PROTO);
-  push<uint8_t>(2);
+  push<uint8_t>(PROTOCOL_VERSION);
 
   // All attributes get pushed into a list and their indices saved in the
   // module def
-  push<OpCode>(OpCode::EMPTY_LIST);
-  push<OpCode>(OpCode::MARK);
+  wrap_in_list_ = wrap_in_list;
+  if (wrap_in_list_) {
+    push<OpCode>(OpCode::EMPTY_LIST);
+    push<OpCode>(OpCode::MARK);
+  }
 }
 
 void Pickler::finish() {
-  push<OpCode>(OpCode::APPENDS);
+  if (wrap_in_list_) {
+    push<OpCode>(OpCode::APPENDS);
+  }
   push<OpCode>(OpCode::STOP);
+
+  // Add the binary data for all the tensors to be included in the same binary
+  // TODO: The pickler should be refactored to stream out to a file directly
+  // instead of staging in the stack_ array
+
+  // Insert storage keys
+  if (literal_tensors_.size() > 0) {
+    Pickler p;
+    p.start(/*wrap_in_list=*/false);
+    auto storage_keys = IValue(c10::ivalue::GenericList::create({}));
+    size_t index = 0;
+    for (auto tensor : literal_tensors_) {
+      storage_keys.toGenericList()->elements().push_back(std::to_string(index));
+      index++;
+    }
+    p.addIValue(storage_keys);
+    p.finish();
+    stack_.insert(
+        stack_.end(), p.stack().data(), p.stack().data() + p.stack().size());
+
+    for (auto tensor : literal_tensors_) {
+      // first dump size
+      auto numel = tensor.numel();
+      auto numel_ptr = reinterpret_cast<const char*>(&numel);
+      stack_.insert(stack_.end(), numel_ptr, numel_ptr + sizeof(numel));
+
+      uint64_t record_size = tensor.element_size() * tensor.numel();
```

Comment:
Is the tensor contiguous? Is the tensor on the CPU? Otherwise this code is bogus.  Look at how the other tensor serializer writes out tensors.